marimo/_server/session/session_view.py

```diff
@@ -0,0 +1,162 @@
+        # Find the first cell that is waiting for stdin.
+        for cell_op in self.cell_operations.values():
+            console_ops: list[CellOutput] = as_list(cell_op.console)
+            for cell_output in console_ops:
+                if cell_output.channel == CellChannel.STDIN:
+                    cell_output.channel = CellChannel.STDOUT
+                    cell_output.data = f"{cell_output.data} {stdin}\n"
+                    return
+
+    def add_operation(self, operation: MessageOperation) -> None:
+        """Add an operation to the session view."""
+
+        if isinstance(operation, CellOp):
+            previous = self.cell_operations.get(operation.cell_id)
+            self.cell_operations[operation.cell_id] = merge_cell_operation(
+                previous, operation
+            )
+        elif isinstance(operation, Variables):
+            self.variable_operations = operation
+
+            # Remove any variable values that are no longer in scope.
+            names: set[str] = set(
+                [v.name for v in self.variable_operations.variables]
+            )
+            next_values: dict[str, VariableValue] = {}
+            for name, value in self.variable_values.items():
+                if name in names:
+                    next_values[name] = value
+            self.variable_values = next_values
+
+        elif isinstance(operation, VariableValues):
+            for value in operation.variables:
+                self.variable_values[value.name] = value
+
+        elif isinstance(operation, Interrupted):
+            # Resolve stdin
+            self.add_stdin("")
+
+    @property
+    def operations(self) -> list[MessageOperation]:
+        all_ops: list[MessageOperation] = [
+            self.variable_operations,
+            VariableValues(variables=list(self.variable_values.values())),
+        ]
+        all_ops.extend(self.cell_operations.values())
+        return all_ops
+
+
+def merge_cell_operation(
+    previous: Optional[CellOp],
+    next_: CellOp,
+) -> CellOp:
+    """Merge two cell operations."""
+    if previous is None:
+        return next_
+
+    assert previous.cell_id == next_.cell_id
+
+    if next_.status is None:
+        next_.status = previous.status
+
+    # If we went from queued to running, clear the console.
+    if next_.status == "running" and previous.status == "queued":
+        next_.console = []
+    else:
+        combined_console: list[CellOutput] = as_list(previous.console)
+        combined_console.extend(as_list(next_.console))
+        next_.console = combined_console
+
+    # If we went from running to running, use the previous timestamp.
+    if next_.status == "running" and previous.status == "running":
+        next_.timestamp = previous.timestamp
+
+    if next_.output is None:
+        next_.output = previous.output
+
+    return next_
+
+
+def as_list(value: Union[Any, Optional[Any], list[Any]]) -> list[Any]:
+    if not value:
```

Comment:
I think this is what you want? `not value` will evaluate to `True` for `False`-y objects

```suggestion
    if value is None:
```